Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
```

Comment:
`CartesianPowerIterator` is not intended to ever provide non-const access to the values referenced by underlying iterator range, and it's not necessary to provide non-const access to the  vector of copies stored in the `CartesianPowerIterator`. The current implementation reduces copying by reusing this vector between iterations, so providing non-const access to it is especially undesirable (https://github.com/Wangscape/Wangscape/pull/64#discussion_r96147811).

I think that considering the intention of the class, it's more appropriate to call it a `const_iterator` than an `iterator`. Note that `string_view` has `typedef const_iterator iterator` ([cppreference](http://en.cppreference.com/w/cpp/string/basic_string_view#Member_types))

I can change its API to make the constness more explicit (specify const `reference` and `pointer` types in `iterator_traits`, make `operator*` return a `reference`), and I can implement more of the `InputIterator` concept. It lacks:
* `operator->`
* `operator++(int)`
* `CopyAssignable`
* `CopyConstructible`